src/module.c

```diff
@@ -1488,35 +1488,95 @@ int RM_ReplyWithSimpleString(RedisModuleCtx *ctx, const char *msg) {
+            addReplyDeferredLen(c);
+        ctx->postponed_arrays_count++;
+    } else {
+        switch (type) {
+        case COLLECTION_REPLY_ARRAY:
+            addReplyArrayLen(c,len);
+            break;
+        case COLLECTION_REPLY_MAP:
+            addReplyMapLen(c,len);
+            break;
+        case COLLECTION_REPLY_SET:
+            addReplySetLen(c,len);
+            break;
+        case COLLECTION_REPLY_ATTRIBUTE:
+            addReplyAttributeLen(c,len);
+            break;
+        case COLLECTION_REPLY_PUSH:
+            addReplyPushLen(c,len);
+            break;
+        default:
+            serverPanic("Invalid module reply type %d", type);
+        }
+    }
+    return REDISMODULE_OK;
+}
+
+
 /* Reply with an array type of 'len' elements. However 'len' other calls
  * to `ReplyWith*` style functions must follow in order to emit the elements
  * of the array.
  *
  * When producing arrays with a number of element that is not known beforehand
  * the function can be called with the special count
- * REDISMODULE_POSTPONED_ARRAY_LEN, and the actual number of elements can be
+ * REDISMODULE_POSTPONED_LEN, and the actual number of elements can be
  * later set with RedisModule_ReplySetArrayLength() (which will set the
  * latest "open" count if there are multiple ones).
  *
  * The function always returns REDISMODULE_OK. */
 int RM_ReplyWithArray(RedisModuleCtx *ctx, long len) {
-    client *c = moduleGetReplyClient(ctx);
-    if (c == NULL) return REDISMODULE_OK;
-    if (len == REDISMODULE_POSTPONED_ARRAY_LEN) {
-        ctx->postponed_arrays = zrealloc(ctx->postponed_arrays,sizeof(void*)*
-                (ctx->postponed_arrays_count+1));
-        ctx->postponed_arrays[ctx->postponed_arrays_count] =
-            addReplyDeferredLen(c);
-        ctx->postponed_arrays_count++;
-    } else {
-        addReplyArrayLen(c,len);
-    }
-    return REDISMODULE_OK;
+    return moduleReplyWithCollection(ctx, len, COLLECTION_REPLY_ARRAY);
+}
+
+/* Very similar to RedisModule_ReplyWithArray except that `len` is
+ * the amount of pairs, so the actual number of elements (i.e. the
+ * number of times to call a `ReplyWith*` style function)
+ * is `len`*2
+ * Visit https://github.com/antirez/RESP3/blob/master/spec.md for more info about RESP3. */
+int RM_ReplyWithMap(RedisModuleCtx *ctx, long len) {
+    return moduleReplyWithCollection(ctx, len, COLLECTION_REPLY_MAP);
+}
+
+/* Very similar to RedisModule_ReplyWithArray
+ * Visit https://github.com/antirez/RESP3/blob/master/spec.md for more info about RESP3. */
+int RM_ReplyWithSet(RedisModuleCtx *ctx, long len) {
+    return moduleReplyWithCollection(ctx, len, COLLECTION_REPLY_SET);
+}
+
+/* Very similar to RedisModule_ReplyWithMap
+ * Visit https://github.com/antirez/RESP3/blob/master/spec.md for more info about RESP3. */
+int RM_ReplyWithAttribute(RedisModuleCtx *ctx, long len) {
+    return moduleReplyWithCollection(ctx, len, COLLECTION_REPLY_ATTRIBUTE);
```

Comment:
I believe this is dangerous, attribute on resp2 is translated to an array that will get you out of sync. Today the only place that uses attribute is `debug protocol attrib` and when used with resp2 you can this:
```
127.0.0.1:6379> debug protocol attrib
1) "key-popularity"
2) 1) "key:123"
   2) (integer) 90
127.0.0.1:6379> ping
"Some real reply following the attribute"
127.0.0.1:6379> hello
PONG
```
It's fine on debugging command, no one should really use it, and it's for debugging purposes only. But a module that will try to use attribute will encounter this problem if the user will use resp2 (and there is no way for a module to know that he should not send the attribute). I see 3 options by which we can handle this:
1. Give the module a way to know if the client is using resp2 or resp3 (might be useful regardless of this issue) and pass the full responsibility to the module (module should not reply with attribute on resp2)
2. Adding a mechanism such that Redis will not write the attribute to a resp2 client, Redis will see that the user is sending an attribute of size 3 (for example) which means that Redis need to ignore the next 6 replies (3 keys + 3 values). This is a very complicated solution taking into account that values can also be nested and might also contain attributes (in theory) ...
3. `RM_ReplyWithAttribute` will return an error if used with resp2, in such case the module should stop sending the attribute content.

Personally, I vote for 3, I think it's a good solution and not too much complicated...

Reply:
Good catch! I prefer 3 too.